hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs

```diff
@@ -461,58 +492,60 @@ data IdeState = IdeState
   -- | A queue of requests to be performed once a module is loaded
   , requestQueue :: Map.Map FilePath [UriCacheResult -> IdeM ()]
   , extensibleState :: !(Map.Map TypeRep Dynamic)
-  , ghcSession  :: Maybe (IORef HscEnv)
+  , ghcSession  :: !(Maybe (IORef HscEnv))
   }
 
 instance MonadMTState IdeState IdeGhcM where
-  readMTS = lift $ lift $ lift readMTS
-  modifyMTS = lift . lift . lift . modifyMTS
-
-instance MonadMTState IdeState IdeDeferM where
   readMTS = lift $ lift readMTS
   modifyMTS = lift . lift . modifyMTS
 
+instance MonadMTState IdeState IdeDeferM where
+  readMTS = lift readMTS
+  modifyMTS = lift . modifyMTS
+
 instance MonadMTState IdeState IdeM where
   readMTS = lift readMTS
   modifyMTS = lift . modifyMTS
 
 class (Monad m) => LiftsToGhc m where
   liftToGhc :: m a -> IdeGhcM a
 
-instance GM.MonadIO IdeDeferM where
-  liftIO = liftIO
+--instance GM.MonadIO IdeDeferM where
+--  liftIO = liftIO
 
 instance LiftsToGhc IdeM where
-  liftToGhc = lift . lift
+  liftToGhc = lift
 
 instance LiftsToGhc IdeGhcM where
   liftToGhc = id
 
 instance HasGhcModuleCache IdeGhcM where
-  getModuleCache = lift $ lift getModuleCache
-  setModuleCache = lift . lift . setModuleCache
+  getModuleCache = lift getModuleCache
+  modifyModuleCache = lift . modifyModuleCache
 
 instance HasGhcModuleCache IdeDeferM where
   getModuleCache = lift getModuleCache
-  setModuleCache = lift . setModuleCache
+  modifyModuleCache = lift . modifyModuleCache
 
 instance HasGhcModuleCache IdeM where
   getModuleCache = do
     tvar <- lift ask
-    state <- liftIO $ readTVarIO tvar
+    state <- readTVarIO tvar
     return (moduleCache state)
-  setModuleCache !mc = do
+  modifyModuleCache f = do
     tvar <- lift ask
-    liftIO $ atomically $ modifyTVar' tvar (\st -> st { moduleCache = mc })
+    atomically $ modifyTVar' tvar (\st -> st { moduleCache = f (moduleCache st) })
```

Comment:
This is just unsafe. Make a ticket that `modifyModuleCache` should be in `STM` monad.